lib/Gocdb_Services/ServiceService.php

```diff
@@ -1798,51 +1798,69 @@ public function editEndpoint(\User $user, \EndpointLocation $endpoint, $newValue
-            $interfaceName = $newValues ['SERVICEENDPOINT'] ['INTERFACENAME'];
-        } else {
-            $interfaceName = ( string ) $service->getServiceType ();
-        }
+        $interfaceName = $newValues ['SERVICEENDPOINT'] ['INTERFACENAME'];
 
         if($newValues['IS_MONITORED']) {
             $monitored = true;
         } else {
             $monitored = false;
         }
 
-        if (empty ( $name )) {
-            throw new \Exception ( "An endpoint must have a name." );
-        }
+        $this->editEndpointLogic($endpoint, $name, $url, $interfaceName, $description, $email, $monitored);
+    }
 
-        // check endpoint's name is unique under the service
-        foreach ( $service->getEndpointLocations () as $endpointL ) {
-            // exclude itself
-            if ($endpoint != $endpointL && $endpointL->getName () == $name) {
-                throw new \Exception ( "Please provide a unique name for this endpoint." );
-            }
-        }
+    public function editEndpointApi (\EndpointLocation $endpoint, $name, $url, $interfaceName, $description, $email, $monitored, $authIdentifierType, $authIdentifier) {
+      //Check the portal is not in read only mode, throws exception if it is
+      $this->checkGOCDBIsNotReadOnly();
 
-        $this->em->getConnection ()->beginTransaction ();
+      \Factory::getSiteService()->checkAuthroisedAPIIDentifier($endpoint->getService()->getParentSite(), $authIdentifier, $authIdentifierType);
+
+      $this->editEndpointLogic($endpoint, $name, $url, $interfaceName, $description, $email, $monitored);
 
-        try {
-            // Set the endpoints new member variables
-            $endpoint->setName ( $name );
-            $endpoint->setUrl ( $url );
-            $endpoint->setInterfaceName ( $interfaceName );
-            $endpoint->setDescription ( $description );
-            $endpoint->setEmail($email);
-            $endpoint->setMonitored($monitored);
-            $this->em->merge ( $endpoint );
-            $this->em->flush ();
-            $this->em->getConnection ()->commit ();
-        } catch ( \Exception $ex ) {
-            $this->em->getConnection ()->rollback ();
-            $this->em->close ();
-            throw $ex;
-        }
     }
 
+    /**
+     *Function containing the logic to edit an endpoint with none of the authorisation or validation
+    */
+    private function editEndpointLogic (\EndpointLocation $endpoint, $name, $url, $interfaceName, $description, $email, $monitored){
+      $service = $endpoint->getService ();
+
+      if (empty ( $name )) {
+          throw new \Exception ( "An endpoint must have a name." );
+      }
+
+      //if no interface name is provided, default to service type
+      if ($interfaceName == '') {
+          $interfaceName = ( string ) $service->getServiceType ();
+      }
+
+      // check endpoint's name is unique under the service
+      foreach ( $service->getEndpointLocations () as $endpointL ) {
```

Comment:
This is done twice?

Reply:
Yep. Here it is for the web portal as well. There it ensures an error message (and response code) on the API that is appropriate.